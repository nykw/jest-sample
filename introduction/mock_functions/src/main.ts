export function forEach<X, F extends (x: X) => unknown>(items: X[], callback: F) {
  for (let index = 0; index < items.length; index++) {
    callback(items[index]);
  }
}
