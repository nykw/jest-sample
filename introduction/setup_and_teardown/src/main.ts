type City = {
  name: string;
  foods: string[];
};
const cities: City[] = [
  { name: 'Vienna', foods: ['Wiener Schnitzel'] },
  { name: 'San Juan', foods: ['Mofongo'] },
];

const isValidCityFoodPair = (name: string, food: string): boolean => {
  const idx = db.cities.findIndex((city) => city.name === name);
  if (idx == -1) return false;
  return db.cities[idx].foods.includes(food);
};

let db = {
  cities,
};

const initializeCityDatabase = async () => {
  db = { cities };
};
const clearCityDatabase = async () => {
  db = { cities };
};

const initializeFoodDatabase = async () => {
  db = { cities };
};

const isCity = async (city: string) => {
  // DBから検索
  return db.cities.map(({ name }) => name).includes(city);
};

export {
  initializeCityDatabase,
  clearCityDatabase,
  isCity,
  isValidCityFoodPair,
  initializeFoodDatabase,
};
