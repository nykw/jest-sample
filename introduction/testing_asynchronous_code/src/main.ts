const fetchData = (callback: (...x: unknown[]) => void) => {
  setTimeout(() => {
    callback('peanut butter');
  }, 100);
};

const fetchDataPromise = () =>
  new Promise<string>((resolve) => {
    setTimeout(() => {
      resolve('peanut butter');
    }, 100);
  });

const fetchDataRejectPromise = async () => {
  throw new Error('error');
};

export { fetchData, fetchDataPromise, fetchDataRejectPromise };
