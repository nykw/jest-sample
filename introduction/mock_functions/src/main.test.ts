import { forEach } from './main';
import foo from './foo';

// see https://jestjs.io/ja/docs/jest-object#jestmockmodulename-factory-options
jest.mock('./foo', () => {
  return {
    __esModule: true,
    default: jest.fn(() => 42),
    foo: jest.fn(() => 42),
  };
});

test('Using a mock function', () => {
  const mockCallback = jest.fn((x) => 42 + x);
  forEach([0, 1], mockCallback);

  // The mock function is called twice
  expect(mockCallback.mock.calls.length).toBe(2);

  // The first argument of the first call to the function was 0
  expect(mockCallback.mock.calls[0][0]).toBe(0);

  // The first argument of the second call to the function was 1
  expect(mockCallback.mock.calls[1][0]).toBe(1);

  // The return value of the first call to the function was 42
  expect(mockCallback.mock.results[0].value).toBe(42);
});

test('Mock Return Values', () => {
  const myMock = jest.fn();
  console.log(myMock());
  // > undefined

  myMock.mockReturnValueOnce(10).mockReturnValueOnce('x').mockReturnValue(true);

  console.log(myMock(), myMock(), myMock(), myMock());
  // > 10, 'x', true, true

  const filterTestFn = jest.fn();

  // Make the mock return `true` for the first call,
  // and `false` for the second call
  filterTestFn.mockReturnValueOnce(true).mockReturnValueOnce(false);

  const result = [11, 12].filter((num) => filterTestFn(num));

  console.log(result);
  // > [11]
  console.log(filterTestFn.mock.calls);
  // > [ [11], [12] ]
});

test('Mock Implementations', () => {
  foo();
  // > 42
  console.log(foo());

  const myMockFn = jest
    .fn()
    .mockImplementationOnce((cb) => cb(null, true))
    .mockImplementationOnce((cb) => cb(null, false));

  myMockFn((err: Error, val: any) => console.log(val));
  // > true

  myMockFn((err: Error, val: any) => console.log(val));
  // > false

  const myMockFn2 = jest
    .fn(() => 'default')
    .mockImplementationOnce(() => 'first call')
    .mockImplementationOnce(() => 'second call');

  console.log(myMockFn2(), myMockFn2(), myMockFn2(), myMockFn2());
  // > 'first call', 'second call', 'default', 'default'
});

test('Mock Names', () => {
  const myMockFn = jest
    .fn()
    .mockReturnValue('default')
    .mockImplementation((scalar) => 42 + scalar)
    .mockName('add42');
  // expect(myMockFn(1)).toBe(42);
});

test('Custom Matchers', () => {
  const mockFunc = jest.fn((...xs: unknown[]) => {});

  const [arg1, arg2] = [1, 2];
  mockFunc();
  mockFunc(arg1, arg2);

  // The mock function was called at least once
  expect(mockFunc).toHaveBeenCalled();

  // The mock function was called at least once with the specified args
  expect(mockFunc).toHaveBeenCalledWith(arg1, arg2);

  // The last call to the mock function was called with the specified args
  expect(mockFunc).toHaveBeenLastCalledWith(arg1, arg2);

  // All calls and the name of the mock is written as a snapshot
  expect(mockFunc).toMatchSnapshot();
});
