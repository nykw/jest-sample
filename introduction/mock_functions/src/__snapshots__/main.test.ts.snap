// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Matchers 1`] = `
[MockFunction] {
  "calls": Array [
    Array [],
    Array [
      1,
      2,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
